### PR TITLE
Pr ekf ecl

### PR DESCRIPTION
**Pull request by tumbili**
This PR brings in the "range aid feature" for the estimator.

To test is set the parameter "EKF2_RNG_AID" to enabled and make sure to test with a vehicle that has a range finder connected.

### CODE REVIEW

**Comment by LorenzMeier** [comment]
@tumbili CI failure.

**Comment by LorenzMeier** [comment]
Jenkins test this please

**Comment by ChristophTobler** [comment]
@tumbili here's a log from a flight I did with a Aero and Lidar Lite V3 http://logs.px4.io/plot_app?log=bb8372b4-60bc-4c7c-94a4-b1cdddf0efb0

**Comment by ChristophTobler** [comment]
rebased the PR

**Comment by tumbili** [comment]
@ChristophTobler I added two parameters for the maximum horizontal velocity and the maximum absolute altitude allowed during range aid mode.

**Comment by ndepal** [comment]
Finally got around to giving this a try. Here are a couple of logs from a 250 quad with a PixFalcon and a Lidar Lite v3.

1) In this flight it seemed to work quite well, pretty nice altitude hold. I was flying over a flat patch of grass. Then, between 80s and 150s in the log, I was flying forward, where the terrain has a ramp up  of approx 1m, and came to a stop where the terrain is once again flat. Then back again, forward again, back again. Then I moved sideways and moved a chair under the drone while at stand still.
One note about the forward-backward maneuver: While the altitude setpoint never changed, the drone lost quite a bit of altitude. This can be seen in the range data. Before moving forward (75s), range data shows height above ground approx 2.7m. After moving forward and back again, the ranger reports approx 2.1m. Yet the height estimate remains constant. I don't know if this has anything to do with the range aid of if this is generally an issue (I don't have a lot of recent experience with EKF2).

screen shot 2017-04-21 at 16 28 45

http://logs.px4.io/plot_app?log=74f847d2-3b3d-4cc0-8bc3-42761db66145

2) This was my first flight with the range aid flag enabled. I saw quite strange behavior with repeated jumps in altitude. The logs show jumps in the altitude setpoints as well as the estimated altitude.
Unfortunately, I did not log ekf2 innovations in this flight. I was not able to reproduce this behavior in subsequent flights.
screen shot 2017-04-21 at 17 31 52

http://logs.px4.io/plot_app?log=a19a8a77-ae1a-410c-b691-30aecfa1451b

**Comment by tumbili** [comment]
@ndepal Thanks Nicolas for the great feedback! I will look at your problems in depth and report back to you. Oh yeah, would be really important if you always log innovations. Thanks!

**Comment by tumbili** [comment]
@ndepal Thanks Nicolas for the great feedback! I will look at your problems in depth and report back to you. Oh yeah, would be really important if you always log innovations. Thanks!

**Comment by LorenzMeier** [comment]
How does this PR relate to https://github.com/PX4/Firmware/pull/7095? Would it make sense to rebase it onto that one?

**Comment by tumbili** [comment]
@ndepal I had some time to look at your logs. I think the reason for the drone loosing altitude without the actual altitude setpoint changing is the fact that the ekf makes use of a height sensor offset variable which is used to make the height measurement consistent with the momentary height state at the time of switching. Imagine you are hovering in one place for some time and only range finder measurements are fused. If you start to move after 10min then the baro has probably already drifted so much that the values are not consistent anymore with the current height state of the filter. To avoid this we use this offset variable.

I think the problem may be that when you are moving the baro is affected by the wind speed and shows a higher altitude. At the instant when the switch back to the range finder happens the estimated height hasn't yet returned to the "true" value and thus the height offset variable is used to compensate for the difference. If you do the same procedure for some time I assume that the offset accumulates and thus, you get the observed effect.

Maybe we need to wait until the baro innovations have settled a bit before we switch to range finder and vice versa.  What do you think?

**Comment by tumbili** [comment]
@LorenzMeier It's not related, but I should still rebase on master.

**Comment by ndepal** [comment]
Thanks for the feedback, @tumbili. Your hypothesis sounds reasonable. I tried to see it in the log (http://logs.px4.io/plot_app?log=74f847d2-3b3d-4cc0-8bc3-42761db66145). Indeed, the baro increases at speed. However, looking at the graph below, it seems that by the time the switch back to the ranger happens (e.g. around 100.5s) the baro has already settled back to steady state (mostly). What I was looking for to support your theory would be that the switch back to ranger fusion (visible by the decrease in z innovation variance) would happen while baro measurements are increased due to velocity.

I noticed something else that is potentially related: The HAGL estimate sees a spike during the horizontal translation. I understand that the raw range measurements see that spike, as the vehicle is tilted and the ranger is not directed to the ground perpendicularly. However, I would assume that the HAGL estimate would account for the tilt.
I have to admit that I did not take advantage of the ranger offset parameters, but as I understand it, an offset of a couple of cm below the axis of rotation would make a rather small difference compared to the effect of the tilt.
The reason I mention this is that the range measurements/HAGL has quite not settled at the switch to ranger at 100.5s.

screen shot 2017-04-26 at 10 27 56

**Comment by tumbili** [comment]
@ndepal  Better to report here
I pushed some more changes which were likely causing your problem. The baro offset calculation which is done when baro is not fused was not correct.
I'm still not happy with how I determine if the terrain is stable to use for range fusion. At the moment I'm just checking the terrain estimator innovations but those can be very noisy and can cause rapid switching between baro and range sensor. I need a better way to determine if the terrain is stable, e.g. by computing a filtered derivative of the terrain state. I'll think about this and report back.

**Comment by ChristophTobler** [comment]
@tumbili @ndepal I think the rangefinder behavior of ekf needs to change. Currently it changes altitude if there is a big change in distance. I don't think it should change altitude if there are abrupt changes, but account for it with a altitude offset. If there's a slope, it should keep the same distance. Generally, a drone should always avoid abrupt changes in any direction.
With that behavior, I don't see why we can not use the "normal rangefinder mode", if it is in a certain range and meets the velocity requirements. Would that be a solution?

**Comment by ndepal** [comment]
I did another two test flights of this today, both times with problematic results. Both times the drone suddenly shot up into the sky while in position mode. It looks like this happens right when horizontal speed is below 0.7 and the switch back to range aid is made. However, it does not happen always when switching back to range aid, as can be seen in the logs.

http://review.px4.io/plot_app?log=ea8424fa-abf4-43c3-af6d-72d1f1a88aa8

http://review.px4.io/plot_app?log=874836ba-fcb4-4160-9b75-704d50e9a80b

Additionally, some comments regarding the flight while it was still stable: I did not fly over the terrain slope in today's second flight (as I did in the flights I reported on April 26), but rather flew sideways/forward and back over a level surface.
A significant increase in altitude was observed every time I stopped translating and the switch to range aid was made. This can be seen in the graph below (of the second log): Every time the horizontal velocity drops, the drone climbs significantly (with an overshoot) and then settles at a higher altitude, as seen in the range measurements. This happens very repeatably until the drone shoots up seemingly indefinitely. Perhaps that last shoot-up, which I aborted by going into manual, is actually the same problem as the other climbs on stopping, just more exaggerated.

screen shot 2017-05-16 at 13 42 11

@tumbili are you able to reproduce this on real hardware?

**Comment by tumbili** [comment]
@ndepal Thanks, something seems to be pretty broken. The switch to baro does not work for some reason. I'll look and try to fix.

**Comment by ndepal** [comment]
What makes you say the switch to baro is broken? At least from the innovation variances, it looks like  it does the switch, right?

**Comment by tumbili** [comment]
@ndepal Yes, but the innovations are very hight after the switch. It might be related to the last commit by @ChristophTobler

**Comment by tumbili** [comment]
@ndepal Yeah, I can reproduce the problem in SITL. I think the problem is this https://github.com/PX4/ecl/pull/244/commits/e4eba908b179277091de60f0cc3b5a963eb78b1d#diff-4cac31c333867233bb8a08de1e812a9aL729

I need to sync with @ChristophTobler on this.

**Comment by ChristophTobler** [comment]
Thanks @ndepal! Hm, strange... I was not yet able to reproduce it in sitl but I'm on it. @tumbili innovations are exactly the same as before -> ~4

**Comment by tumbili** [comment]
@ChristophTobler Hm, from a quick glance at the last commit is seems that you are not zeroing the height sensor offset when switching back to baro fusion. This offset is related to the range finder and causes the high baro innovations. Correct me if I'm wrong but I don't see what could justify the code which zeros the offset.

**Comment by ndepal** [comment]
I'm trying to understand the purpose of `_hgt_sensor_offset` and `_baro_hgt_offset`. 
@tumbili Is it true that you need that second offset variable `_baro_hgt_offset` only because you want to keep changing it while using the range sensor or the GPS for height, for which `_hgt_sensor_offset` needs to be constant?
Am I also correct in assuming that `_hgt_sensor_offset` must always be 0 here: https://github.com/PX4/ecl/blob/bcffef38b3d02ce0d98381499160a5a692cdf6ae/EKF/vel_pos_fusion.cpp#L138 ?
Lastly, is it true that `_baro_hgt_offset` is not being changed while baro is used for height fusion?

If so, would it not be better to change the line https://github.com/PX4/ecl/blob/bcffef38b3d02ce0d98381499160a5a692cdf6ae/EKF/control.cpp#L730 to

    _hgt_sensor_offset = _baro_hgt_offset;

and removing `_hgt_sensor_offset` from https://github.com/PX4/ecl/blob/bcffef38b3d02ce0d98381499160a5a692cdf6ae/EKF/vel_pos_fusion.cpp#L138 ?

**Comment by tumbili** [comment]
@ndepal Yes, all of your assumptions are correct. I think your suggestion makes sense, I'll change it.
@ChristophTobler is just taking care of the problem mentioned before, we were able to reproduce it both in SITL finally.

**Comment by ndepal** [comment]
I just did two more test flights with ecl reverted to bcffef. I observed some more strange behavior: While keeping the z setpoint constant, the drone would slowly but continuously climb. This happened in two subsequent flights. I'm not sure this has anything to do with this PR; do I have a problem with my setup (sensor noise or something)?

http://logs.px4.io/plot_app?log=b78e7d43-bc23-4fbf-97c6-cdedb7ddd17a
http://logs.px4.io/plot_app?log=de433a5c-c44e-4d19-a778-58e6a3f53546

**Comment by ChristophTobler** [comment]
@ndepal I can reproduce the same in SITL. I think it's because the switch between the two modes. The error adds up over time.

**Comment by tumbili** [comment]
@ndepal Thansk I'll look at them as well.

**Comment by ndepal** [comment]
@ChristophTobler if you're referring to the last two logs I posted, I don't think it came from the switches. It was very continuous, it looked as if I didn't have my throttle stick centered.
The logs show horizontal velocity always below 1, so no switches should be occurring. Weirdly though, since horizontal velocity is always low, it should just be in range mode, but the z innov var suggest otherwise.

**Comment by ChristophTobler** [comment]
@ndepal Yes, you're right, I just had a look at the log. From the description, it sounded like another problem which is the drone gaining altitude with every switch.

**Comment by ndepal** [comment]
Since I haven't flown this drone with EKF2, perhaps a baseline test of a well tested EKF2 version is in order. If I remember correctly this pr (in ecl) has a whole lot of commits that are not in master yet. What would be the closest tested version of EKF2?

**Comment by ndepal** [comment]
Two more test flights with ecl on bcffef.

First flight: same parameters as previous logs, once again observed slow and steady climb.
http://logs.px4.io/plot_app?log=7a6b1fd4-3f1c-4e5e-a648-ed7e979bee58

Second flight: EKF_RNG_AID set to 0 for comparison. While there altitude variations of around 0.5m during hover, no significant drift was observed.
http://logs.px4.io/plot_app?log=8797e396-d1ee-4f96-8fde-d583eae7debb

Thus, the slow and steady climb is likely related to this feature, even though no switching is occuring in the first log. The z innovation variance indicates baro fusion, even though the range aid conditions are met.

**Comment by tumbili** [comment]
@ndepal Thanks for the testing! Something is still not right, because it does not switch anymore to range and at some point it does not fuse baro anymore, so no height is being fused and eventually a reset happens. Now you know how good the accel is for predicting the height :-) I'll debug this after work and test on hardware.

**Comment by mhkabir** [comment]
@tumbili @ChristophTobler Please let me know when this is ready for testing.

**Comment by tumbili** [comment]
@mhkabir This is not ready and I need some time on the weekend to debug it. I can now reproduce the same problem as @ndepal I'll let you know.

**Comment by ChristophTobler** [comment]
@tumbili I just had a go on a snapdragon with the current state. I couldn't reproduce @ndepal 's behavior... Seems to have worked just fine. We can still improve on recovering altitude after switches though.
range_aid_snapdragon
http://logs.px4.io/plot_app?log=5aaf6dbb-14ef-4036-9ea5-dc4cfe97552f

**Comment by tumbili** [comment]
@ChristophTobler Thanks, I'll have a look at the logs in the evening.

**Comment by LorenzMeier** [comment]
@tumbili Any updates on this?

**Comment by ChristophTobler** [comment]
rebased and added a commit to match https://github.com/PX4/ecl/commit/871ebb251b15af3745559c8d46bd8e7fde6df107

File: src/modules/ekf2/ekf2_params.c [hunk 1]
```diff
@@ -994,3 +994,40 @@ PARAM_DEFINE_FLOAT(EKF2_MAGB_VREF, 2.5E-7f);
...
+ *
+ * If this parameter is enabled then the estimator will make use of the range finder measurements
+ * to estimate it's height even if range sensor is not the primary height source. It will only do so if conditions
+ * for range measurement fusion are met.
+ *
+ * @group EKF2
+ * @value 0 Range aid disabled
+ * @value 1 Range aid enabled
+ */
+PARAM_DEFINE_INT32(EKF2_RNG_AID, 0.0f);
```
Comment (priseborough) [inline src/modules/ekf2/ekf2_params.c]: nitpick - should be 0, not 0.0f

Comment (ChristophTobler) [inline src/modules/ekf2/ekf2_params.c]: You're right, Thanks!

**Comment by priseborough** [comment]
I rebased on latest ecl master and test flew over a range of speeds and terrain that included large ramps and steps using an early model LIDAR-lite sensor that has limited range of about 6m and frequent measurement dropouts.  HMAX was set to 3 for this test which was a reasonable test of how this algorithm will cope with a cheaper grade sensor than the current generation LL40LS  The height keeping when stationary was good. Repeated start stop manoeuvres at the start of flight would see the height ratchet up a by about 0.1m at each stop, however this characteristic disappered when the accel bias had been learned .

[06_59_25.ulg.pdf](https://github.com/PX4/Firmware/files/1080007/06_59_25.ulg.pdf)

**Comment by tumbili** [comment]
@priseborough Thanks for testing! Could you share entire log please?

**Comment by bkueng** [comment]
I've successfully test-flown this, here are some logs with different param settings:
- EKF2_RNG_AID=1, EKF2_HGT_MODE=0: http://logs.px4.io/plot_app?log=5de92888-aaa0-4d14-b510-e2865f92e643
- EKF2_RNG_AID=1, EKF2_HGT_MODE=1: http://logs.px4.io/plot_app?log=a1482b33-3274-4254-98f7-35cc07d8b7fc
- EKF2_RNG_AID=1, EKF2_HGT_MODE=2: http://logs.px4.io/plot_app?log=34bbd043-a5f2-424e-a222-817a37fa07cd
- EKF2_RNG_AID=1, EKF2_HGT_MODE=1, EKF2_RNG_A_HMAX=30.0: http://logs.px4.io/plot_app?log=19e61a10-11fe-4c16-81d6-9f7b3c00823d

@tumbili can we increase the maximum of `EKF2_RNG_A_HMAX` to eg. 100m (and maybe add a comment that this is sensor-specific)? I have a range finder with ~40m range and I assume future sensors will become even better. Or do you have a reason to keep it that small?

@dagar When I flashed this, the `EKF2_RNG_AID` was not available (neither in QGC nor with `param set`, and also not in `src/modules/systemlib/param/px4_parameters.h`), but they existed in `parameters.xml` in the build directory. After a clean build they were available on the vehicle as well. Is there a cmake dependency missing?

**Comment by dagar** [comment]
I'll check the parameter dependencies.

**Comment by priseborough** [comment]
The problem with large range thresholds at the moment is that the height estimate becomes increasingly sensitive to terrain and roll/pitch estimation errors which can lead to range sensor innovation check failures and timeouts. The ecl EKF has an internal parameter which enables the scaling of error with range to be allowed for, but it is not exposed to the user in ekf2_main. I have a patch here which adds the parameter.

https://github.com/priseborough/Firmware/commit/ee6b87659023bfe459dd57276af4406b61897e34

**Comment by ChristophTobler** [comment]
Here are two more logs
* EKF2_RNG_AID=1, EKF2_HGT_MODE=0, EKF2_RNG_A_VMAX = 1 m/s: http://logs.px4.io/plot_app?log=3456a354-da20-44a3-bb8b-64b334b85c3f
* EKF2_RNG_AID=1, EKF2_HGT_MODE=0, EKF2_RNG_A_VMAX = 5 m/s: http://logs.px4.io/plot_app?log=67b27584-9c2e-4f15-9ba9-d64ab0429e92

Works like a charm! I think we can get this in

**Comment by priseborough** [comment]
I have tested this with a copter equipped with optical flow and can confirm that it worked well in that application. See: http://logs.px4.io/plot_app?log=071cd922-2823-4eb6-8076-3cc54c8a7def

@tumbili This can be brought in when the ecl version patch is updated.

**Comment by tumbili** [comment]
@priseborough Ok, doing this now.

**Comment by tumbili** [comment]
@priseborough I rebased this on master and cherry-picked those last two commits of yours.
If the checks pass we can merge it.
Thanks @priseborough @bkueng @ChristophTobler @ndepal @dagar @LorenzMeier for the support!

**Comment by ndepal** [comment]
Have any changes been made that address the problems I was having?
If not, is there something I can do (tuning, damping, ...) to make this work for me? Based on the tests I have done, my impression is that this is unreliable at best.

**Comment by tumbili** [comment]
@ndepal I can compare your last log with the newest logs posted here and check if I can find spot a difference.
@priseborough said 

> Repeated start stop manoeuvres at the start of flight would see the height ratchet up a by about 0.1m at each stop, however this characteristic disappered when the accel bias had been learned

which may be related to your observations.

**Comment by tumbili** [comment]
@ndepal I can't guarantee that I haven't done any changes since your last flight, I honestly lost track of that. So if you have the bandwidth it would for sure not hurt if you could test latest master with this feature so that we are sure we are comparing apples with apples.

**Comment by ChristophTobler** [comment]
@ndepal Are you sure you have updated the submodules? The behavior you described sounds like the issue I introduced in ecl. Or can you maybe retest with current master?

**Comment by ndepal** [comment]
@tumbili I can probably do another test this week. But it does seem to me that there definitely needs to be some kind of safeguarding against the situations that I encountered where apparently no sensor is used for height updates (i.e. open loop accel integration).

**Comment by ndepal** [comment]
@ChristophTobler I'm sure. If you look at my comments above, e.g. https://github.com/PX4/Firmware/pull/6989#issuecomment-301800132 and https://github.com/PX4/Firmware/pull/6989#issuecomment-301821794, I specifically tested a commit of ecl that did not have your issue (that commit has no longer exists now, as ecl has since been rebased)

**Comment by tumbili** [comment]
@ndepal Ok, I went through the history and see what you mean. I'll come back to you!

**Comment by priseborough** [comment]
@ndepal  I've checked the 7a6b1fd4 log where the ratcheting behaviour was reported and it appears that the range finder data was never accepted by the EKF for fusion so that when it switches to range finder fusion there is no data there ind only the GPS vertical velocity is constraining drift. Was this setup working previously on master with EKF2_HGT_MODE=2 ? If not recommend that that you do this first.

**Comment by tumbili** [comment]
@ndepal @priseborough I had a look at your log 7a6b1fd4 and may have found a bug in the logic.
These are my observations:
1) The ekf never switched to range mode, although range data was available (you can check the ekf control status flag which confirms that it was always using baro).
2) Although the system should have been fusion barometer measurements it did not do so for long periods. You can see that the height innovations are constant over long periods which means that no height was  being fused. As a result the ekf reset it's height several times as can be seen when observing the z reset counter in the local position topic.
3) At some points in the log you can see by looking at the height innovation that baro measurements have been fused again, see e.g. at T = 26s.

My assumption is that the filter can get into a faulty state if the range finder has been declared faulty (_rng_hgt_faulty) and we are using the range aid feature. This [line](https://github.com/PX4/ecl/blob/master/EKF/control.cpp#L723) shows that the _fuse_height flag will not be set if range aid conditions are met but we have a faulty range finder. In that case no height measurement will be fused and the innovations will be constant as in the log. The fact that at some points in the log baro measurements have been fused from time to time can be explained by the fact that range aid conditions were not met anymore. This can be confirmed by observing large horizontal movements in the log.
So I assume we either need to guard against entering range aid mode if we have a faulty range sensor or we need to allow using baro if range aid conditions are met but the range sensor is faulty.

I'll have to have a closer look again but his is what I found out so far.

**Comment by ndepal** [comment]
I flashed master today and did a few test flights.

EKF_HGT_MODE 2:
Terrain following worked with the exception of two instances where I flew over an obstacle (table), causing a jump in range measurements. I was expecting the quad to jump up too, but it appears the range sensor was just not being fused anymore. The quad descended until it was sitting on the table. This happened both times. I realize that this problem has nothing to do with this new feature.
screen shot 2017-06-22 at 11 50 55_terrain_following
http://logs.px4.io/plot_app?log=2564808a-d9d6-4262-9138-4c281180ac4c

EKF2_HGT_MODE 0 EKF2_RNG_AID 1:
Range aid with baro as the main source worked for me. I did not notice too much ratcheting.
screen shot 2017-06-22 at 11 52 41_baro_rng_aid
http://logs.px4.io/plot_app?log=0122a5a7-6b82-4ebe-a64a-1ee4781740b2

EKF2_HGT_MODE 1 EKF2_RNG_AID 1:
Range aid with GPS as the main source did not appear to work. After moving quickly and then stopping, the quad's height had quite significant overshoot/undershoot before settling. If I'm interpreting the logs right, it never actually switched to range fusion.
screen shot 2017-06-22 at 12 02 48_gps_rng_aid
http://logs.px4.io/plot_app?log=3d65c43d-3b3b-421c-a9f8-bfa32d7ff402


What went wrong with the last flight? Why did it never fuse range?

What is the problem with the terrain following flight? I understand that range measurements are initially rejected after flying over the table, but it seems to me the measurements are rejected for too long.

**Comment by ChristophTobler** [comment]
@ndepal If the z innovations are to big (relative to the variance) there will be a reset and it does not follow the terrain. You could adjust (increase) `EKF2_RNG_GATE` so that it will follow higher gaps.

**Comment by ChristophTobler** [comment]
@tumbili I think you are right with the `_rng_hgt_faulty` assumption. I assume the problem was that `_rng_hgt_faulty` was true when the `EKF2_HGT_MODE` was switched to baro (from range finder). Because `_rng_hgt_faulty` can only be false again if you are using the range finder mode (https://github.com/PX4/ecl/blob/7f1cca5b28087055f8343ec069852bedd3b86cd7/EKF/control.cpp#L624). So we need to protect against that.
So if that happens range aid mode is not working because of https://github.com/PX4/ecl/blob/7f1cca5b28087055f8343ec069852bedd3b86cd7/EKF/control.cpp#L723 as @tumbili mentioned. I think the last statement (`!_in_range_aid_mode`) is not needed.

**Comment by tumbili** [comment]
@ChristophTobler Yes, I agree. Still need to find out if this is the same reason why in the last log of @ndepal it never switched to range finder when he was using it together with gps altitude.

**Comment by ChristophTobler** [comment]
Here are some fixes https://github.com/PX4/ecl/pull/286. @tumbili would be good if you could have a look at it.

Could be that the problem @ndepal had when using it with GPS was that GPS was faulty and never changed back to false because the flag was set to rng_hgt all the time. (the ekf status flag is always on rng_hgt)

**Comment by tumbili** [comment]
@ChristophTobler Thanks, I'll have a look.

**Pull request merged**